source/multigrid/mg_transfer_prebuilt.cc

```diff
@@ -85,15 +85,9 @@ MGTransferPrebuilt<VectorType>::prolongate(const unsigned int to_level,
   Assert((to_level >= 1) && (to_level <= prolongation_matrices.size()),
          ExcIndexRange(to_level, 1, prolongation_matrices.size() + 1));
 
-#ifdef DEBUG
-  if (this->mg_constrained_dofs != nullptr)
-    Assert(this->mg_constrained_dofs->get_user_constraint_matrix(to_level - 1)
-               .get_local_lines()
-               .size() == 0,
-           ExcNotImplemented());
-#endif
-
   prolongation_matrices[to_level - 1]->vmult(dst, src);
```

Comment:
We found when implementing for MGTransferMatrixFree (in pull request https://github.com/dealii/dealii/pull/8730) that we need to distribute the user constraints before prolongation (i.e. moving `this->mg_constrained_dofs->get_user_constraint_matrix(to_level).distribute()` to before this `->vmult()` call), and then do nothing for the restriction.

Reply:
I tried out your proposed changes. I am testing these changes on my multigrid for the Eddy Current problem. I am testing with 2,3,4,5 global refinements and with a multiplicative Schwarz smoother. 
NOTATION: {n_global_refinements} -> {n_iteration_steps}

1. No distribute() in restrict, distribute() **before** prolongation: 2 -> 21, 3 -> 25, 4 -> 26, 5 -> KILLED

2. No distribute() in restrict, distribute() **after** prolongation: 2 -> 5, 3 -> 6, 4 -> 6, 5 -> 6

3. without your proposed changes: 2 -> 5, 3 -> 6, 4 -> 5, 5 -> 5, 6 -> 5, 7 -> 5

Runs 1 - 3 only differ by the changes to the deal.ii library.

Reply:
This is very curious to me since applying constraints after prolongation should not have an effect. Maybe I am missing a difference in MGTransferPrebuilt that is not in MGTransferMatrixFree. 

This may be silly, but could you do the run but without distributing constraints at all during the transfer? (i.e. just commenting out the Assert that doesn't allow for user constraints and not calling distribute).

Reply:
I created this pull request, because beforehand I didn't distribute any constraints at all during the transfer. My results were: 2 -> 21, 3 -> 25, 4 -> 26, 5 -> 30. Obviously the issue then was the lack of the constraints in the transfer

Reply:
Hmm, yeah, this confuses me as to why this is the case. So distributing only before prolongation iterations are the exact same as if using no constraints. @tjhei thoughts?

Clearly your iteration counts are low and constant, so this appears to be correct for this problem. Let me do some tests with the MatrixFreeTransfer to see if I learn anything new since the two are going to need to be consistent.

Reply:
> So distributing only before prolongation iterations are the exact same as if using no constraints. 

Exactly. And with no constraints the preconditioned system matrix had many eigenvalues close to 0 which indicates that it was singular.

> Let me do some tests with the MatrixFreeTransfer to see if I learn anything new since the two are going to need to be consistent.

Sounds good!